### PR TITLE
Fix multi-physics simulations with "many-to-many" interfaces between markers and improve associated communications.

### PR DESCRIPTION
**Pull request by pcarruscag**
## Proposed Changes
I overlooked something in #951, in that PR I made possible defining many-to-many multi-physics interfaces, i.e. one marker can receive some variables from zone A (e.g. temperature) and others from zone B (e.g. displacements).

The problem is that unless the case is very very trivial, this does not work because currently the target geometry can only store donor information (interpolation coefficients and donor point indices) for one donor.
So many-to-many only works if everything matches, interpolation method, point indices, etc.

The proposed solution is to move the donor information out of CVertex and into the CInterpolator object, which has a one-to-one relation with the interfaces, hopefully this will make the interfaces generic enough.

## Related Work
#951 

## PR Checklist
- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [X] My contribution is commented and consistent with SU2 style.

### CODE REVIEW

File: SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp [hunk 1]
```diff
@@ -247,11 +243,6 @@ void CDiscAdjMultizoneDriver::Run() {
 
       config_container[iZone]->Set_StartTime(SU2_MPI::Wtime());
 
-      if (retape) {
-        SetRecording(NONE, Kind_Tape::FULL_TAPE, ZONE_0);
-        SetRecording(SOLUTION_VARIABLES, Kind_Tape::ZONE_SPECIFIC_TAPE, iZone);
-      }
```
Comment (pcarruscag) [inline SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:253]: @oleburghardt I got rid of the zone-specific recording, the secondary recording still records all zones at once so the peak memory is the same and this way we have one less feature to maintain / debug.

File: Common/include/interface_interpolation/CInterpolator.hpp [hunk 2]
```diff
@@ -78,14 +78,30 @@ class CInterpolator {
   CGeometry* const target_geometry;  /*! \brief Target geometry. */
 
 public:
+  struct CDonorInfo {
```
Comment (oleburghardt) [inline Common/include/interface_interpolation/CInterpolator.hpp:81]: Ok nice, that's neat :-)

File: SU2_CFD/src/interfaces/CInterface.cpp [hunk 3]
```diff
@@ -56,240 +59,144 @@ CInterface::~CInterface(void) {
...
-      /*--- We have identified the local index of the Structural marker ---*/
-      /*--- We loop over all the vertices in that marker and in that particular processor ---*/
+    for (auto iVertex = 0ul; iVertex < target_geometry->GetnVertex(markTarget); iVertex++) {
+      const auto iPoint = target_geometry->vertex[markTarget][iVertex]->GetNode();
 
-      for (iVertex = 0; iVertex < target_geometry->GetnVertex(Marker_Target); iVertex++) {
+      if (!target_geometry->nodes->GetDomain(iPoint)) continue;
 
-        Point_Target = target_geometry->vertex[Marker_Target][iVertex]->GetNode();
+      auto& targetVertex = interpolator.targetVertices[markTarget][iVertex];
```
Comment (oleburghardt) [inline SU2_CFD/src/interfaces/CInterface.cpp:165]: Also just from looking at it, this looks way more appropriate than having it stored in the geometry.

**Review by oleburghardt — APPROVED** [review]
Hi Pedro, nice and clean work, thanks. Looks good to me.

Had a closer look at `CIsoparametric` and just skimmed through the other interpolation methods. As for the interface side and from what I can judge, the changes in `BroadcastData` all look good. So from my side, go ahead...

File: SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp [hunk 4]
```diff
@@ -515,10 +505,15 @@ void CDiscAdjMultizoneDriver::SetRecording(unsigned short kind_recording, Kind_T
 
   if ((tape_type == Kind_Tape::OBJECTIVE_FUNCTION_TAPE) || (kind_recording == MESH_COORDS)) {
     HandleDataTransfer();
+    for (iZone = 0; iZone < nZone; iZone++) {
+      if (Has_Deformation(iZone)) {
+        iteration_container[iZone][INST_0]->SetDependencies(solver_container, geometry_container, numerics_container,
+                                                            config_container, iZone, INST_0, kind_recording);
+      }
+    }
+    SetObjFunction(kind_recording);
```
[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:514]: Yup, for that small case the gradients agree super well with finite differences. For larger scale FSI only problems the agreement is also better, although not perfect yet.

File: SU2_CFD/include/solvers/CEulerSolver.hpp [hunk 5]
```diff
@@ -374,6 +374,7 @@ class CEulerSolver : public CFVMFlowSolverBase<CEulerVariable, COMPRESSIBLE> {
    */
   inline virtual void Viscous_Residual(unsigned long iEdge, CGeometry *geometry, CSolver **solver_container,
                                        CNumerics *numerics, CConfig *config) { }
+  using CSolver::Viscous_Residual; /*--- Silence warning ---*/
```
[the remark this answers is not in the document]
Comment (pcarruscag) [reply SU2_CFD/include/solvers/CEulerSolver.hpp:377]: The Viscous_Residual just above the using declaration has a different signature from the "normal" Viscous_Residual, and so the compiler warns that the former shadows / hides the latter.
The using declaration is the way to tell the compiler that we are doing this on purpose, and not by mistake.

**Comment by pcarruscag** [comment]
Thank you for the review Ole.

File: SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (oleburghardt) [reply SU2_CFD/src/drivers/CDiscAdjMultizoneDriver.cpp:514]: Great! :+1:  Good to know.

File: SU2_CFD/include/solvers/CEulerSolver.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (oleburghardt) [reply SU2_CFD/include/solvers/CEulerSolver.hpp:377]: I see, thanks for the explanation.

**Pull request merged**